### PR TITLE
CLIP-ViT Model Bring up

### LINKED ISSUE #30870: [Bounty $1000] CLIP-ViT bring up using TTNN APIs
### 📝 Background

This bounty is for bringing up CLIP-ViT (Contrastive Language–Image Pre-training using Vision Transformer) using TTNN APIs on Tenstorrent hardware (Wormhole or Blackhole). [CLIP](https://github.com/openai/CLIP), developed by OpenAI, jointly trains an image encoder (ViT) and a text encoder (Transformer) to align their embeddings in a shared latent space.
It enables zero-shot classification and image–text retrieval by comparing cosine similarity between encoded image and text pairs.

The goal of this bounty is to reproduce CLIP-ViT inference on TT hardware — bringing up the ViT encoder for image embedding and text encoder for text embedding — then computing similarity efficiently on-chip.

### 🎯 What Success Looks Like

A successful submission will fulfill all requirements in the following stages. Payout is made after **all three stages** are completed.

### Stage 1 — Bring-Up

- Implement CLIP-ViT using TTNN APIs (Python).
- Model runs on either N150 or N300 hardware with no errors.
- Successfully encodes image and text inputs into embeddings.
- Accuracy evaluated on COCO dataset following [Yolov4 demo instructions](https://github.com/tenstorrent/tt-metal/tree/main/models/demos/yolov4#coco-2017-dataset).
- Clear instructions for setup and running the model.

### Stage 2 — Basic Optimizations

- Apply sharding / interleaving strategies to ViT encoder blocks.
- Fuse linear + activation + layernorm where applicable.
- Store intermediates in L1 where beneficial.
- Use recommended TTNN/tt-metal CNN flows.
- Use TT library of fused ops like [relu with conv](https://github.com/tenstorrent/tt-metal/pull/29236)

### Stage 3 — Deeper Optimization

- Maximize core parallelism across Transformer blocks.
- Implement deeper TT-specific optimizations (e.g., fused QKV projections + SDPA in attention, GEMM+Bias+GELU fusion in MLPs, pre/post-LayerNorm fusion, and layout choices that eliminate transposes/reshapes to avoid unnecessary Tensor manipulations).
- Tune tile sizes / tensor layouts for ViT multi-head attention.
- Document any advanced tuning or known issues.

### 🧭 Guidance & Starting Points

- Use the [TTNN model bring-up tech report](https://github.com/tenstorrent/tt-metal/blob/main/tech_reports/ttnn/TTNN-model-bringup.md) as your primary reference.
- Reference [vit implmentation](https://github.com/tenstorrent/tt-metal/tree/main/models/demos/wormhole/vit)
- Reference [vit documentation](https://github.com/tenstorrent/tt-metal/blob/main/tech_reports/ViT-TTNN/vit.md)
- Reference [Yolov4 implementation (PR 29157)](https://github.com/tenstorrent/tt-metal/pull/29157) for example end-to-end flows.
- Refer to [TT fused ops](https://github.com/tenstorrent/tt-metal/pull/29236) for opportunities to optimize.
- For accuracy validation, reuse ImageNet zero-shot labels from OpenAI’s benchmark.
- Focus on ViT-B/32 or ViT-B/16 variant for the first milestone.
- Ask for help or file issues if ops are missing in TTNN.

### 🔎 Possible Approaches

- Start from an existing PyTorch CLIP-ViT model and port layers one by one to TTNN.  
- Validate each submodule's output (patch embedding, Transformer blocks, text encoder) against the PyTorch reference before full integration.  
- Experiment with different sharding strategies and memory configurations for attention and MLP layers.  
- Use TTNN profiling tools to identify bottlenecks (e.g., attention, projection, or GELU) and areas for fusion.  
- Open a draft PR early to get feedback on your approach.

### 📊 Result submission guidelines notes

Beyond the model implementation itself. Contributors must submit the following material as a proof of work. However, feel free to open a PR at any time if you want us checking you are on the right track. Just understand that payout is only made after all 3 stages are completed.

- Functional model implementation
- Validation logs (output correctness)
- Performance report + header for final review

Links:
- [Link to perf sheet](https://github.com/tenstorrent/tt-metal/blob/main/tech_reports/ttnn/TTNN-model-bringup.md#41-performance-sheet)
- [Perf header](https://docs.tenstorrent.com/tt-metal/latest/ttnn/ttnn/profiling_ttnn_operations.html#perf-report-headers)

---

### 📚 Resources

- [TTNN model bring-up tech report](https://github.com/tenstorrent/tt-metal/blob/main/tech_reports/ttnn/TTNN-model-bringup.md)
- [CNN Bring-up & Optimization in TT-NN](https://github.com/tenstorrent/tt-metal/blob/main/tech_reports/CNNs/cnn_optimizations.md)
- [Yolov4 TTNN implementation PR](https://github.com/tenstorrent/tt-metal/pull/29157)
- [OpenAI CLIP GitHub Repo](https://github.com/openai/CLIP)
- [TT fused ops PR](https://github.com/tenstorrent/tt-metal/pull/29236)
- [Perf report header](https://docs.tenstorrent.com/tt-metal/latest/ttnn/ttnn/profiling_ttnn_operations.html#perf-report-headers)

### PR DESCRIPTION
**Pull request by ajumpa**
### Ticket
#30870 

### Problem description
- This PR includes model bring up of CLIP-ViT vision-language model on TT n300s hardware. 
- I use a block sharding strategy to maximize throughput (>2x from baseline implementation) by retaining activations in L1.
- This strategy enables much better Tensix core utilization, eliminates extra permutations/reshapes, and kernel/data movement dispatches.
- Other optimizations include the use of memconfigs, program configs, activation fusion, and qkv fusion. 
- Tested and developed with an n300s with 8x7 core grid, batch sizes are thus best in multiples of 7 -> [7,14,21,...35,...56,63]

### Files

tests
- tests/test_pcc.py
- tests/test_e2e_throughput.py

demo
- demo.py

tt (model)
- tt/tt_clip_model.py
- tt/tt_clip_text.py
- tt/tt_clip_vision.py
- tt/tt_clip_model_optimized.py
- tt/tt_clip_text_optimized.py
- tt/tt_clip_vision_optimized.py

Please see README.md for details on running tests and demo

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1212832436210754

### CODE REVIEW

**Comment by ajumpa** [comment]
@jberkowitzTT polishing up a couple things. This PR will be ready for review by Monday afternoon, the 13th

**Comment by jberkowitzTT** [comment]
I will have to reopen this tomorrow if this isn't complete by then

Thanks for the hard work on this

**Comment by ajumpa** [comment]
@jberkowitzTT My PR is ready for review!

**Comment by ajumpa** [comment]
@jberkowitzTT @yieldthought @uaydonat please let me know if I must submit any changes. I can do so tonight

**Comment by ajumpa** [comment]
@jberkowitzTT 84 checks have passed, looks like some have skipped?

**Comment by yieldthought** [comment]
- `pytest models/demos/clip_vit/demo.py -sv` matched HuggingFace top-1 on 7/7 images.
- `pytest models/demos/clip_vit/demo.py -sv --opt --b16` matched HuggingFace top-1 on 6/7 images.
- `pytest models/demos/clip_vit/demo.py -sv --opt --b8` matched HuggingFace top-1 on 2/7 images.
- `pytest models/demos/clip_vit/tests/test_e2e_throughput.py -sv -k 280` reported `172.5` images/s for base bf16 and `435.6` images/s for optimized bf16, a `2.525x` speedup at `N=280`.

I note that in the parent issue Stage 1 asked for a PR that includes COCO/Yolov4-style accuracy evaluation. I'm not sure checking 7 images it really enough but will have an agent look at the rest.

**Comment by yieldthought** [comment]
Running repeated batches causes massive accuracy loss on the second batch. If you close the device and reopen it between batches then you get good results, but this bug needs to be fixed before the model can be used in practice.

COCO val2017 zero-shot category evaluation over 4,952 annotated images (reopening the device between each batch):
- optimized bf16 scored top1 0.81 and top5 0.96
- optimized bfloat8_b scored top1 0.81 and top5 0.96
- HF reference scored 0.86 top-1 and 0.97 top-5
I think these are ok, but the multi-batch corruption bug needs to be addressed.

**Comment by yieldthought** [comment]
Update: I think the multi-batch corruption is a TTNN bug. I will address it separately to this PR.

**Review by yieldthought — APPROVED** [review]

**Comment by mickg10** [comment]
**The multi-batch corruption does not reproduce in an independent, non-sharded CLIP ViT implementation — but I first produced a convincing false positive, which may be worth ruling out here too.**

Context: I independently wrote a ttnn CLIP **ViT-L/14** vision tower (functional style — interleaved DRAM, `ttnn.linear` + fused `scaled_dot_product_attention`, no block sharding, no custom program configs) running on 4× Wormhole n300 (8 chips) as one `MeshShape(1,8)`, batch sharded with replicated weights, tt-metal `e95ffa59`. Since this PR is blocked on repeated-batch corruption, I tried to reproduce it.

### First attempt reproduced the symptom exactly — and it was my own bug

Scoring 2000 images as 8 consecutive 256-image batches in a single device session (never reopened), compared against a torch fp32 reference with identical preprocessing:

```
                      before fix          after fix
run-to-run differing  98 / 2000           0 / 2000
batch 0 vs torch      0.0590 (max 0.20)   0.0590 (max 0.20)
batch 1 vs torch      0.2108 (max 2.15)   0.0614 (max 0.29)
```

Batch 0 bit-exact, batch 1 onward degraded and non-deterministic between runs — i.e. exactly the reported signature.

It was **not** ttnn. It was host-side: my shared-memory staging slots are addressed by index, but were guarded by a *counting* semaphore whose permits are fungible. A worker that had run ahead to image 1200 could take a permit freed by image 0 and overwrite slot 176 while the main thread had yet to read slot 176. Fixed by staging in waves so a slot is never rewritten while its data is still live.

Flagging it because it is an easy way to manufacture a "second batch is corrupted" result that looks like a device bug. If the failing setup here stages input through any reused host buffer, ring, or pinned-memory pool, it may be worth eliminating that before concluding it is TTNN.

### After that fix, it does not reproduce

- 2000 images, 8 consecutive batches, one device session: **0/2000 scores differ between runs**, and batch 1 accuracy (mean |Δ| 0.0614 vs torch fp32) matches batch 0 (0.0590). No second-batch cliff.
- A stricter check, alternating two genuinely different inputs A/B/A/B for 8 iterations without reopening the device (repeating one input can't distinguish "correct" from "reused a stale buffer that happens to hold identical data"):

```
iter  input   max|dev from first of same input|
   0      A                        0.000000e+00
   1      B                        0.000000e+00
   2      A                        0.000000e+00
   ...
   7      B                        0.000000e+00

A vs B differ by (must be non-zero): 2.281250e+00
A cosine vs torch ref (min over 8): 0.984945
```

Bit-exact repeats, and A/B genuinely differ, so nothing is being masked.

### What this narrows down

The structural difference from this PR is that I do **not** use block sharding or L1-resident activations — everything is interleaved DRAM with generic matmul configs. `ttnn.linear`, `scaled_dot_product_attention`, `layer_norm`, `softmax` and the mesh scatter/gather all repeat cleanly across batches here. That points at buffer/L1 reuse in the **sharded** path rather than a general TTNN bug — specifically whether sharded activation buffers are fully re-initialised, rather than assumed-overwritten, on the second and subsequent invocations.

Caveats, plainly: different config (ViT-L/14, 24×1024, 257 tokens padded to 288, not ViT-B), different memory layout, and possibly a different commit than this PR was tested against. So this is evidence about where the bug *isn't*, not a refutation of the report.

Happy to help isolate it — I can run your `tt/tt_clip_vision.py` unchanged on 8× Wormhole against the same harness, or run any specific batch sequence you'd like. I can also share the ~80-line repro harness if useful.
